Display status of item locations by status priority.

GrothTobias opened the pull request:
Currently, if the item status for multiple locations is displayed in search results (Item_Status > multiple_locations = 'group'), locations with unknown items are always displayed as 'unknown' regardless if there are available items or not.

With this PR the displayed location status is based on the status priority configured in AvailabilityStatus:getPriority().

'Uncertain' and 'unavailable' are treated the same, as it is in the current implementation. (Not 'available' and not 'unknown')

TODO
- [ ]  Update changelog when merging (changes to return values of \VuFind\ILS\Logic::availabilityAsString(); significant rewrite of \VuFind\AjaxHandler\GetItemStatuses; simplification of check_item_statuses.js to do less HTML generation).

GrothTobias committed:
Display status of item locations by status priority.

GrothTobias committed:
Removed unneeded code

demiankatz commented:
Thanks, @GrothTobias! Since @ThoWagen designed this system, I'd be interested to hear his thoughts on these changes. Also, do we need to adjust integration tests to help cover this case?

At the moment I'm very busy finishing up some database-related refactoring for release 10, but if either of you think there's a strong value in getting this change included in release 10.0, I'll try to give it a closer look as soon as time permits!

ThoWagen commented on module/VuFind/src/VuFind/AjaxHandler/GetItemStatuses.php:
```diff
@@ -362,16 +352,23 @@ protected function getItemStatusGroup($record, $callnumberSetting)
+
             $locationInfo = [
-                'availability' => $details['available'] ?? false,
+                'availability' =>
```
The test `VuFindTest/Mink/HoldingsTest.php` actually expects the "uncertain" availability to be handled like as "available". This could be fixed by using `$locationAvailability->isAvailable()` here instead.

But I thinks we could refactored this even more and also have the "uncertain" availability handled by its own. For example we could return `$locationAvailability->availabilityAsString()` and remove `status_unknown`. Then we can display the location's availability here https://github.com/vufind-org/vufind/blob/ad39f269eb57607b634b0fd83336a7db59004728/themes/bootstrap3/js/check_item_statuses.js#L62 depending on that string.

What do you think about that?

ThoWagen reviewed (changes requested):
I think the general approach in this PR is very good. Only the "uncertain" case needs some kind of change. See the comment below.

GrothTobias committed:
Refactored display of item locations in the results

GrothTobias commented:
I refactored it a bit to @ThoWagen's approach.

Based on the returned availability, all four statuses can be displayed now.
For now the 'uncertain' icon is the same as 'unknown', if you have ideas or preferences as to which icon to use, let me know.

While I'd like to somewhat separate the classes from the JavaScript, so people can easily change them, I don't really like how I did it here.

As for the integration tests, I'm unsure as to how to change them to reflect the changes.

@ThoWagen, I changed the return values in `AvailabilityStatus:availabilityAsString()` from 'true'|'false'|'unknown'|'uncertain' to 'available'|'unavailable'|'unknown'|'uncertain'. This makes it simpler to handle the response and I think it is more consistent. From what I see the return value was only really used in the section I changed. If I overlooked some usages please let me know so I can change it.

demiankatz commented on themes/bootstrap3/js/check_item_statuses.js:
```diff
@@ -58,27 +58,12 @@ VuFind.register('itemStatuses', function ItemStatuses() {
-            + result.locationList[x].location
-            + '</span> ';
-        }
+        locationListHTML += '<span class="' + availabilityClasses[status] + '">'
```
You should add availabilityClasses to the global declaration at the top of the file to prevent warnings here.

demiankatz reviewed (changes requested):
See below for one small suggestion.

I like the idea of changing 'true' and 'false' to more descriptive terms, but that probably counts as a breaking change that might require us to reschedule this PR from 10.1 to 11.0. Is that a problem for you? Or maybe @ThoWagen feels differently... or maybe there's a way to incorporate some backward compatibility via a deprecated option.

Regarding integration tests, if things are failing after you and @ThoWagen agree on the final shape of the code, I don't mind helping to get the tests to pass.

ThoWagen commented:
Regarding the classes in JS, I think it is good that you are importing them in JS, even though I understand that this does not look like the cleanest way. But currently I can not think of a better way to do this.

And I agree with @demiankatz on the change  from 'true'|'false'|'unknown'|'uncertain' to 'available'|'unavailable'|'unknown'|'uncertain'. This should be part of a major release or done in a backward compatible way.

GrothTobias committed:
Reverted return values of AvailabilityStatus::availabilityAsString. Adjusted check_item_statuses.js.

GrothTobias commented:
Sorry, had to do something else first. 

I reverted the `AvailabilityStatus:availabilityAsString()` to `false | true | uncertain | unavailable` and adjusted the JS since I couldn't think of a good way to make it backward compatible.

The return values would have to be changed in another PR.

GrothTobias committed:
Code styling

GrothTobias committed:
Code styling

GrothTobias committed:
Fixed results-scripts.phtml for theme bootstrap3

demiankatz commented:
Thanks, @GrothTobias! I've asked @ThoWagen to re-review when he has a moment. If he has no concerns with this revision, I'll give it a closer look myself and then (if ready) merge it.

ThoWagen commented on themes/bootstrap3/js/check_item_statuses.js:
```diff
@@ -58,27 +58,17 @@ VuFind.register('itemStatuses', function ItemStatuses() {
       el.querySelectorAll('.callnumber,.hideIfDetailed,.location').forEach((e) => e.classList.add('hidden'));
       var locationListHTML = "";
       for (var x = 0; x < result.locationList.length; x++) {
+        var status = result.locationList[x].status;
```
I believe this is causing the error. `.status` is undefined. You probably want to check for `.availability`.

ThoWagen commented on themes/bootstrap3/js/check_item_statuses.js:
```diff
@@ -58,27 +58,17 @@ VuFind.register('itemStatuses', function ItemStatuses() {
       for (var x = 0; x < result.locationList.length; x++) {
+        var status = result.locationList[x].status;
+        var locationClass =
+          availabilityClasses[status] === 'true' ? 'available'
```
The `availabilityClasses` keys are the "availabilityAsString" values. So `availabilityClasses[availability]` should be sufficient. For the status `availability === 'true'` and `availability === 'false'` need to be checked.

ThoWagen reviewed (changes requested):
Sorry, I just realized that a review from me is still being requested here.

Unfortunately, something is not working properly. The Mink HoldingTest is failing. I think I found the problem. See the comments below.

GrothTobias committed:
Fixed availability display

GrothTobias commented:
Sorry for the late fix.
I don't know why I committed it like that, it couldn't work...

demiankatz commented:
@GrothTobias, thanks for the fix, and sorry for my own slow response. I think @ThoWagen will look at this soon, but he was out of office last week, and I haven't had a chance to look at this yet in his absence. :-)

ThoWagen commented:
Unfortunately there are still two Mink tests failing. But the reason is not some wrong logic but that the tests are not correct anymore. The case when "multiple_locations" is "group" and the availability is "uncertain" or "unknown" needs to be modified. So basically the case "uncertain" needs to be handled somewhere in  the lines 139-163 of the HoldingTest.php and if availability is null in line 174 the type is "muted" rather than "warning".

demiankatz commented:
Thanks for taking another look, @ThoWagen. @GrothTobias, please let me know if you need help fixing the tests and/or getting a standard test environment set up. I'm happy to help if you need me to, but I'll wait to hear from you before taking action in case you prefer to work on this yourself.

GrothTobias commented:
I can't get the vagrant testing environment running. At first I could not import the marc records, that seemed to be a problem with symlinks in a shared folder (sadly, my host system is windows).

Now I have created a new folder and did a `git checkout` and `composer install`. The httpd-vufind.conf and environment variabls were updated with the new vufind location. 
With this I can import the marc data but have a new problem when running `./phing.sh startup`: 
`Error connecting to solr at URL http://localhost:8983/solr/biblio/update : Error connecting to solr server for ping http://localhost:8983/solr/biblio`

demiankatz commented:
@GrothTobias, I confess that I have never tried using the test suite in the Vagrant environment; I typically set up an Ubuntu VM in VirtualBox for testing, as this makes everything a bit more accessible. If that's too much trouble, though, I'm willing to investigate this further in my existing environment and see if I can easily fix the tests. :-)

GrothTobias commented:
If it helps, with a wget on `http://localhost:8983/solr/admin/cores?action=STATUS` I this response:

`{
  "responseHeader":{
    "status":0,
    "QTime":1
  },
  "initFailures":{ },
  "status":{
    "authority":{
      "name":"authority",
      "instanceDir":"/vufind_fork/solr/vufind/authority",
      "dataDir":"data/",
      "config":"solrconfig.xml",
      "schema":"schema.xml",
      "isLoaded":"false"
    },
    "biblio":{
      "name":"biblio",
      "instanceDir":"/vufind_fork/solr/vufind/biblio",
      "dataDir":"/vufind_fork/solr/vufind/biblio/",
      "config":"solrconfig.xml",
      "schema":"schema.xml",
      "startTime":"2024-09-04T13:46:38.632Z",
      "uptime":7709334,
      "index":{
        "numDocs":10,
        "maxDoc":10,
        "deletedDocs":0,
        "version":6,
        "segmentCount":1,
        "current":true,
        "hasDeletions":false,
        "directory":"org.apache.lucene.store.NRTCachingDirectory:NRTCachingDirectory(MMapDirectory@/vufind_fork/solr/vufind/biblio/index lockFactory=org.apache.lucene.store.NativeFSLockFactory@20f0e4ce; maxCacheMB=48.0 maxMergeSizeMB=4.0)",
        "segmentsFile":"segments_2",
        "segmentsFileSizeInBytes":220,
        "userData":{
          "commitCommandVer":"1809066741041463296",
          "commitTimeMSec":"1725260487596"
        },
        "lastModified":"2024-09-02T07:01:27.596Z",
        "sizeInBytes":56004,
        "size":"54.69 KB"
      }
    },
    "reserves":{
      "name":"reserves",
      "instanceDir":"/vufind_fork/solr/vufind/reserves",
      "dataDir":"data/",
      "config":"solrconfig.xml",
      "schema":"schema.xml",
      "isLoaded":"false"
    },
    "website":{
      "name":"website",
      "instanceDir":"/vufind_fork/solr/vufind/website",
      "dataDir":"data/",
      "config":"solrconfig.xml",
      "schema":"schema.xml",
      "isLoaded":"false"
    }
  }
}`

demiankatz commented:
That looks like a healthy Solr response. I suspect the problem is just that the automated testing code wasn't designed to run inside Vagrant. There's probably a way to make it work, but it may require a significant amount of configuration work. Before going down that road, I'll see if I can just solve the test problems in my own environment when time permits. (I unfortunately ran out of time today, but I might have better luck tomorrow!).

demiankatz committed:
Merge branch 'dev' into item-status-location-by-priority

demiankatz commented:
@GrothTobias, I finally found time to run tests and found that there are two failing tests. Here's what the browser looks like during the first failure:

![image](https://github.com/user-attachments/assets/2e6ff448-dada-4775-a9ae-7621441dc069)

That doesn't look quite right! Any idea what might be wrong?

The test that was running had set "multiple_locations" to "group" in config.ini, and had turned on the Demo driver with this configuration:

```
[StaticHoldings]
testsample1 = "[{\"barcode\":\"12345678\",\"availability\":0,\"status\":\"Foo\",\"location\":\"Test Location\",\"locationhref\":false,\"reserve\":\"N\",\"callnumber\":\"Test Call Number\",\"duedate\":\"\",\"is_holdable\":true,\"addLink\":true,\"addStorageRetrievalRequestLink\":\"check\",\"addILLRequestLink\":\"check\",\"__electronic__\":false},{\"barcode\":\"12345678\",\"availability\":0,\"status\":\"Foo\",\"location\":\"main\",\"locationhref\":false,\"reserve\":\"N\",\"callnumber\":\"Test Call Number\",\"duedate\":\"\",\"is_holdable\":true,\"addLink\":true,\"addStorageRetrievalRequestLink\":\"check\",\"addILLRequestLink\":\"check\",\"__electronic__\":false},{\"barcode\":\"12345678\",\"availability\":2,\"status\":\"Check with Staff\",\"location\":\"main\",\"locationhref\":false,\"reserve\":\"N\",\"callnumber\":\"Test Call Number\",\"duedate\":\"\",\"is_holdable\":true,\"addLink\":true,\"addStorageRetrievalRequestLink\":\"check\",\"addILLRequestLink\":\"check\",\"__electronic__\":false}]"
```

If you have a test instance that doesn't have a record with an ID of testsample1, you should still be able to reproduce this scenario -- just change testsample1 in your Demo.ini to a different record ID.

Is that any help? Please let me know if you need me to do further investigation on my end, or if you're having trouble reproducing the scenario.

ThoWagen commented:
@demiankatz @GrothTobias I belive as I said here https://github.com/vufind-org/vufind/pull/3758#issuecomment-2301885492 that the tests are failing because they are not correct anymore. The tests complain about the uncertain's status class "text-warning". It expects "text-succsess". But that is not correct.

However, we can see another problem in Demian's screenshot. The icon is not displayed because it was not yet added to the list in `js-icons.phtml`.

Please let me know if I can help implementing some of the fixes.

GrothTobias committed:
Added 'status-uncertain' icons to js-icons.phtml

GrothTobias commented:
I have added the icons to `js-icons.phtml` (still the same icon as `status-unknown`). 

@ThoWagen It would be great if you could change the tests. As I have said before, i can't get them running.

As a side note, I would probably change the priorities to `0 = uncertain; 1 = unknown; 2 = unavailable; 3 = available` since i think having definite status should be ranked higher than an uncertain one. However, that is probably personal preference, would be a breaking change and can be easily changed by extending `ILS/Logic/AvailabilityStatus.php`.

From my understanding `uncertain` is when the ILS does not returns a status and `unknown` is for things like missing books. If it is the other way, I would swtich both.

EreMaijala commented:
> From my understanding `uncertain` is when the ILS does not returns a status and `unknown` is for things like missing books. If it is the other way, I would swtich both.

"Uncertain" is meant for situations where the item is known to exist, but there's no information on whether the user can have it (could be something like user group restrictions, holdings missing specific item data or such). Unknown is the "more unavailable" one of the two. The current order is for back-compatibility, but isn't necessarily that important anymore if the class and constants are used everywhere.

demiankatz committed:
Fix broken tests.

demiankatz commented on themes/bootstrap3/js/check_item_statuses.js:
```diff
@@ -61,27 +61,17 @@ VuFind.register('itemStatuses', function ItemStatuses() {
-            + result.locationList[x].location
-            + '</span> ';
-        }
+        locationListHTML += '<span class="' + availabilityClasses[availability] + '">'
```
Should we have fallback behavior if `availabilityClasses[availability]` is undefined for some reason (either the AJAX handler returns an unexpected value, or the scripts setting the underlying data get broken)? It might be helpful to log something to the console, or at least avoid a Javascript error caused by an illegal array access.

demiankatz replied:
Or alternatively, would it make sense to just pass the class as part of the AJAX response, and move some of the logic currently added to results-scripts.phtml into GetItemStatuses.php instead?

GrothTobias replied:
I think returning the class and icon in GetItemStatuses.php would be cleaner and is easier to customize. Also the whole availabilityClasses would be unnecessary (I wasn't really happy with how I added it).
We could also directly return the rendered HTML. We have customized the HTML a bit, so returning rendered HTML would be appreciated. :-)
@ThoWagen what do you think?

Regarding the priority: For now I would leave it as is. I think this is a somewhat individual desicion for the libraries, so maybe a little survey would be good to determine what should be used as default.

Also, if we move this to 11.0, I would change the `true `and `false `in `Helper\Root\AvailabilityStatus:availabilityAsString()` to more descriptive terms. Previously I reverted that change in case you wanted to include this PR in 10.1

ThoWagen replied:
@GrothTobias I just remembered that we had this discussion already here https://github.com/vufind-org/vufind/pull/3585#pullrequestreview-2001180237 . So yes, I think moving the logic from JS to GetItemStatuses.php would make sense.

GrothTobias replied:
I wasn't clear if server-side rendering of the location list is wanted, so the classes and icons for each location are now returned in `GetItemStatuses.php`. I added the icon names (not sure what to call it) to `Helper\Root\AvailabilityStatus.php`, similar to the classes. 
With this the field `availability` is not needed for locations but I'm unsure if I should remove it - it could be useful if someone wants to customize this in the JS.

demiankatz replied:
I think at this point I've gone ahead and moved this to the 11.0 milestone, so you can feel free to make larger changes if you like. I don't mind doing more server-side rendering in the AJAX handler as long as it properly accounts for the theme system (e.g. use a template to render where appropriate). Less Javascript code constructing HTML is probably a good thing, for readability and customization. :-)

I'm going to hold off on further review/testing until you're done making the changes you wish to make, so please let me know when you are ready for me to take a closer look again!

GrothTobias replied:
The `GetItemStatuses` handler now returns rendered HTML. As mentioned in the `@TODO` there is still something to do but I don't have time for it next week.

demiankatz replied:
Thanks for the progress; since this is now scheduled for release 11, there's no huge rush to finish -- just please let me know when you're ready for another review.

demiankatz reviewed (changes requested):
Thanks, @GrothTobias and @ThoWagen -- I've adjusted the tests so they pass.

I also gave the code another review and had a couple of thoughts -- see below.

I also feel like this might be best moved to the 11.0 milestone since it does make some significant architectural changes. If you feel otherwise, I'm happy to discuss further, though! :-)

Finally, please let me know when you're satisfied with the order of precedence; I wasn't entirely sure from recent conversation whether that should or should not be adjusted further.

GrothTobias committed:
Return location list's icons and classes in ajax response.

GrothTobias committed:
Removed unused 'availabilityClasses' declaration from check_item_statuses.js

GrothTobias committed:
Server side rendering of location list and callnumbers.

GrothTobias committed:
Fixed code style

GrothTobias committed:
Fixed code styling

GrothTobias committed:
code style fix

demiankatz commented on Vagrantfile:
```diff
@@ -55,7 +55,7 @@ Vagrant.configure("2") do |config|
     mkdir -p /vufindlocal/cache/cli /vufindlocal/config/vufind
     chown -R vagrant:vagrant /vufindlocal
-    su - vagrant -c 'cd /vagrant && composer install && php install.php --non-interactive --overridedir=/vufindlocal'
+    su - vagrant -c 'cd /usr/local/vufind/vufind_fork && composer install && php install.php --non-interactive --overridedir=/vufindlocal'
```
Were these changes to the Vagrantfile meant to be committed?

GrothTobias replied:
I'm very sorry, this wasn't intended to be comitted. I reverted it.

demiankatz commented on config/vufind/Demo.ini:
```diff
@@ -91,8 +91,9 @@ services[] = 'custom'
-;[StaticHoldings]
+[StaticHoldings]
 ;1234 = '[{"location": "foo"}, {"location": "bar", "use_status_class": true, "extraStatusInformation": {"extraInfo": true}}]'
+testsample1 = "[{\"barcode\":\"12345678\",\"availability\":0,\"status\":\"Foo\",\"location\":\"Test Location\",\"locationhref\":false,\"reserve\":\"N\",\"callnumber\":\"Test Call Number\",\"duedate\":\"\",\"is_holdable\":true,\"addLink\":true,\"addStorageRetrievalRequestLink\":\"check\",\"addILLRequestLink\":\"check\",\"__electronic__\":false},{\"barcode\":\"12345678\",\"availability\":0,\"status ...
```
This also looks like an accidental commit...

GrothTobias replied:
I'm very sorry, this wasn't intended to be comitted. I reverted it.

GrothTobias committed:
Revert unwanted changes

GrothTobias committed:
Refactored pickValue() to return an array. Use associative array for callnumbers instead of a formatted string.

GrothTobias committed:
Added changes to bootstrap3 theme. Fixed code styling.

GrothTobias committed:
Fixed code styling.

GrothTobias committed:
Fixed code styling.

GrothTobias commented:
@demiankatz I have refactored the code so it does not need to create strings and explode the callnumbers and results from `GetItemStatuses::pickValue()`. 

While I'm not quite satisfied due to serializing the callnumber arrays to use `array_unique` in `GetItemStatuses::155`, I can't think of a good way to do this besides going back to creating a callnumber string and exploding it when needed.

Do you have an idea to do this in a better way?

demiankatz commented on Vagrantfile:
```diff
@@ -27,7 +27,7 @@ Vagrant.configure("2") do |config|
   config.vm.network :forwarded_port, guest: 80, host: 4567
   config.vm.network :forwarded_port, guest: 8983, host: 4568
-  config.vm.synced_folder ".", "/vagrant", :owner => 'vagrant'
+   config.vm.synced_folder ".", "/vagrant", :owner => 'vagrant'
```
The changes to this file look like unintentional commits.

demiankatz commented on themes/bootstrap3/templates/ajax/itemLocationList.phtml:
```diff
@@ -0,0 +1,12 @@
+  <div class="groupCallnumber">
+    <?= $location['callnumber']?>
+  </div>
+<?php endforeach; ?>
```
It's probably worth adding an end-of-file line break here for style consistency.

demiankatz commented on themes/bootstrap3/templates/ajax/itemCallnumber.phtml:
```diff
@@ -0,0 +1,5 @@
+<?php if ($this->callnumberHandler): ?>
+  <a href="<?= $this->url('alphabrowse-home')?>?source=<?= $this->escapeUrl($this->callnumberHandler)?>&from=<?= $this->escapeUrl($this->actualCallnumber)?>"><?= $this->escapeHtml($this->displayCallnumber)?></a>
+<?php else: ?>
+  <?= $this->displayCallnumber?>
```
Should probably add an escapeHtml() here.

demiankatz commented on themes/bootstrap3/templates/ajax/itemLocationList.phtml:
```diff
@@ -0,0 +1,12 @@
+    </span>
+  </div>
+  <div class="groupCallnumber">
+    <?= $location['callnumber']?>
```
Should probably add an escapeHtml() here.

demiankatz commented on module/VuFind/src/VuFind/AjaxHandler/GetItemStatuses.php:
```diff
@@ -127,7 +130,7 @@ protected function translateList($transPrefix, $list)
         $transList = [];
         foreach ($list as $current) {
-            $transList[] = $this->translateWithPrefix($transPrefix, $current);
+            $transList[] = is_string($current) ? $this->translateWithPrefix($transPrefix, $current) : $current;
```
Maybe worth a comment here: what is `$current` when it is not a string? (e.g. something like `// $current may be a call number array instead of a string, in which case we should bypass translation`).

GrothTobias replied:
I will add a comment. In the current VuFind code `pickValue` is not called with a prefix when used for callnumbers, and so `translateList` will not be called. 
I added that mostly to be safe in case there are institutions that want to translate their callnumbers with prefixes.

demiankatz commented on module/VuFind/src/VuFind/AjaxHandler/GetItemStatuses.php:
```diff
@@ -219,17 +213,48 @@ protected function reduceServices(array $rawServices)
+            }
+
+            $html[] = $this->renderer->render('ajax/itemCallnumber', [
+                'actualCallnumber' => $actualCallnumber,
```
You could use `compact('actualCallnumber', 'displayCallnumber', 'callnumberHandler')` to make this code a little simpler.

demiankatz reviewed (changes requested):
@GrothTobias, thanks for the progress here. I've given this another review and have a few more comments.

I can't think of a better way to deduplicate the lists; I don't expect that these should ever be so unreasonably huge that this operation will be too expensive.

GrothTobias committed:
Added missing escapeHtml calls in templates. Code style update.

GrothTobias committed:
Vagrantfile fix

demiankatz commented on module/VuFind/src/VuFind/AjaxHandler/GetItemStatuses.php:
```diff
@@ -127,7 +130,8 @@ protected function translateList($transPrefix, $list)
         $transList = [];
         foreach ($list as $current) {
-            $transList[] = $this->translateWithPrefix($transPrefix, $current);
+            // $current can be an array if pickValues() is called with callnumbers and a translation prefix
```
Typo: wrong method name.

```suggestion
            // $current can be an array if pickValue() is called with callnumbers and a translation prefix
```

demiankatz commented on module/VuFind/src/VuFind/AjaxHandler/GetItemStatuses.php:
```diff
@@ -219,17 +214,47 @@ protected function reduceServices(array $rawServices)
+
+            $html[] = $this->renderer->render(
+                'ajax/itemCallnumber',
+                compact($actualCallnumber, $displayCallnumber, $callnumberHandler)
```
Syntax is wrong (though an easy mistake to make, since this is a weird function); should be:

```suggestion
                compact('actualCallnumber', 'displayCallnumber', 'callnumberHandler')
```

GrothTobias replied:
I fixed this.

On another note: I just created a new branch from upstream/dev but it's already 98 commits ahead without doing anything. How can this be?

demiankatz replied:
This PR is now targeted against dev-11.0 rather than dev. I'm trying to merge dev into dev-11.0 frequently, but is it possible the anomaly you're seeing is related to this?

GrothTobias replied:
My error. I should have synced the dev branch in my fork...

GrothTobias replied:
That was about a new branch i created for another possible PR. The problem was that the dev branch in my fork wasn't up-to-date and i created the branch from upstream. It's alright now.

If I check out the upstream's dev branch, I get an error after logging in - `Laminas\ServiceManager\Exception\ServiceNotFoundException : Unable to resolve service "VuFind\Role\PermissionProvider\PluginManager" to a factory; are you certain you provided it during configuration?`
It seems the error exists since merging #3882 or is it just me?

demiankatz replied:
Does clearing out your $VUFIND_LOCAL_DIR/cache directory make a difference? Sometimes old cached configurations can cause this kind of problem. If that doesn't help, let me know and I'll see if I can get this branch running in my own environment.

GrothTobias replied:
Thanks! Now it works!

demiankatz reviewed (changes requested):
Thanks, @GrothTobias -- see below for a couple of follow-up suggestions.

It also looks like you need to run `vendor/bin/phing update-bootstrap5` to get the themes back in sync.

GrothTobias committed:
Syntax fix

GrothTobias committed:
Sync bs3 and bs5 themes

GrothTobias committed:
Merge remote-tracking branch 'refs/remotes/upstream/dev' into item-status-location-by-priority

demiankatz committed:
Merge branch 'dev-11.0' into item-status-location-by-priority

demiankatz commented:
@GrothTobias, the extra files were showing here because this branch was behind the current dev-11.0 branch. I've merged the latest changes in, and now the appropriate diffs seem to be showing again.

I'm interested in whether @ThoWagen has further comments on this latest version. I'll give him a chance to comment and will take another look myself after that.

ThoWagen commented on module/VuFind/src/VuFind/AjaxHandler/GetItemStatuses.php:
```diff
@@ -127,7 +130,8 @@ protected function translateList($transPrefix, $list)
         $transList = [];
         foreach ($list as $current) {
-            $transList[] = $this->translateWithPrefix($transPrefix, $current);
+            // $current can be an array if pickValue() is called with callnumbers and a translation prefix
```
Small inaccuracy in the comment. `$current` is always an array when called with callnumbers even if there is not translation prefix.

GrothTobias replied:
I'm sorry if this caused confusion. Maybe I just don't know how to word it correctly.

`translateList` will not be called for callnumbers without translation prefixes since `pickValue()` would just return the list:
`return $transPrefix ? $this->translateList($transPrefix, $list) : $list;`

ThoWagen replied:
Okay, my bad. I didn't notice that. I believe you could still remove the " and a translation prefix" part. But since this is not wrong, I don't mind leaving this as is.

ThoWagen commented on themes/bootstrap3/templates/ajax/itemLocationList.phtml:
```diff
@@ -0,0 +1,12 @@
+    </span>
+  </div>
+  <div class="groupCallnumber">
+    <?=$this->escapeHtml($location['callnumber'])?>
```
`$location['callnumber']` can be HTML if a callnumberHandler is used. So this should not be escaped here.
(This is causing the failure of the CallnumberBrowseTest)

GrothTobias replied:
I fixed this.

ThoWagen reviewed (commented):
Besides the two comments below this is looking good to me!

GrothTobias committed:
Fixed callnumbers in location lists

GrothTobias committed:
Fixed callnumbers in location lists

ThoWagen commented on module/VuFind/src/VuFind/AjaxHandler/GetItemStatuses.php:
```diff
@@ -140,36 +144,27 @@ protected function translateList($transPrefix, $list)
      * @param string $msg         Message to display if $mode == "msg"
      * @param string $transPrefix Translator prefix to apply to values (false to
-     * omit translation of values)
+     *                            omit translation of values)
```
This line break is not needed anyway. So you could also remove it.

ThoWagen reviewed (commented):
Besides that small suggestion this looks good to me :)

GrothTobias committed:
Updated PHPDoc and comment

ThoWagen reviewed (approved):
Looking good to me now!

demiankatz commented on themes/bootstrap5/templates/ajax/itemLocationList.phtml:
```diff
@@ -0,0 +1,12 @@
+<?php $availabilityStatus = $this->availabilityStatus(); ?>
+<?php foreach ($this->locationList as $location): ?>
+  <div class="groupLocation">
+    <span class="<?= $availabilityStatus->getClass($location['availability'])?>">
```
Should we `escapeHtmlAttr` here just in case getClass returns something weird?

GrothTobias replied:
I added the escape function.

demiankatz commented on themes/bootstrap5/templates/ajax/itemLocationList.phtml:
```diff
@@ -0,0 +1,12 @@
+    </span>
+  </div>
+  <div class="groupCallnumber">
+    <?=$location['callnumber']?>
```
This needs to be escaped.

GrothTobias replied:
This is already rendered HTML from `itemCallnumber.phtml`, see `GetItemStatuses.php:374` and `GetItemStatuses.php:238-256`. I changed the name for less confusion in the class, check_item_statuses.js and template.

demiankatz commented on themes/bootstrap5/templates/ajax/itemCallnumber.phtml:
```diff
@@ -0,0 +1,5 @@
+<?php if ($this->callnumberHandler): ?>
+  <a href="<?= $this->url('alphabrowse-home')?>?source=<?=$this->escapeUrl($this->callnumberHandler)?>&from=<?=$this->escapeUrl($this->actualCallnumber)?>"><?=$this->escapeHtml($this->displayCallnumber)?></a>
```
I wonder if it would be a little more readable (and slightly more safe) if you built the href URL into a variable in a first step, and then called `escapeHtmlAttr` on it to build the `a` tag.

Also, there's nothing wrong with using `$this->escapeUrl()`, but for whatever reason, we don't use that helper anywhere else in the code, instead calling `urlencode` directly. If you prefer the helper, I don't object, but if it's all the same, maybe it's better to change to `urlencode` for consistency (and if we want to use the helper, we could consider replacing all the `urlencode` calls with helper calls in a separate PR).

GrothTobias replied:
I changed the `escapeUrl`/`urlencode `to `http_build_query` as i think it has better readability and used a variable.

demiankatz replied:
Thanks, that definitely makes things more readable!

demiankatz reviewed (changes requested):
@GrothTobias, thanks for all the progress here (and thanks for taking time to review, @ThoWagen).

I think we're nearly done, but I had a few minor comments/suggestions related to the templates, mostly related to being paranoid about escaping things. I realize that some of the missing escaping has been faithfully ported from the previous Javascript, but we might as well fix them regardless if they need fixing. (And if any data is pre-escaped and I'm being overly paranoid, feel free to ignore any irrelevant suggestions).

GrothTobias committed:
Renamed callnumber variable. Escape html attributes.

demiankatz reviewed (approved):
Thanks again, @GrothTobias -- this looks good to me and all tests are passing!
